From 730b4f769072b878d2737ba0b57fab67dacbd9e1 Mon Sep 17 00:00:00 2001
From: Lei wang <quic_leiwan@quicinc.com>
Date: Fri, 14 Apr 2023 10:52:05 +0800
Subject: [PATCH] systemd: Reduce max sessions id to save dentry consumpition

systemd will create session-$(session-id).scope file when user login, kernel
will create related dentry as well. Continusly login/logout will create huge
bunch of reclaimable dentry slab.
Only in high memory pressure, kernel will prefer reclaim slab as shrink slab
need take long time, it will cause process hung in alloc memory, furthermore
it may lead to oom issue.
As session-max is 8192 by default, tune the max session id to session-max * 8 to
reduce extra dentry consumption.

Upstream-Status: Pending

Signed-off-by: Lei wang <quic_leiwan@quicinc.com>
---
 src/login/logind-dbus.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/login/logind-dbus.c b/src/login/logind-dbus.c
index b3174e4b8d..ea06607910 100644
--- a/src/login/logind-dbus.c
+++ b/src/login/logind-dbus.c
@@ -859,6 +859,9 @@ static int method_create_session(sd_bus_message *message, void *userdata, sd_bus
                 do {
                         id = mfree(id);
 
+                        if (_likely_(m->sessions_max !=0 && (UINT64_MAX / m->sessions_max) > (1<<3)))
+                                m->session_counter = m->session_counter >= (m->sessions_max << 3) ? 0 : m->session_counter;
+
                         if (asprintf(&id, "c%lu", ++m->session_counter) < 0)
                                 return -ENOMEM;
 
-- 
2.37.3

